File: 15_3Sum/solution.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        /*
            Input: array of integers nums
            Output: vector of vector of integers, the triplets that sum to 0. Elements must not all come from the same index

            Sample test case
            1.) nums = [-1,0,1,2,-1,-4],
                output = [[-1,-1,2],[-1,0,1]]
            
            Strategy
            1.) Brute force O(n3)
                - sort input array
                - triple for loop computing the sum of triplets and shortlisting those that sum to 0 into a set to remove duplicates
                std::vector<std::vector<int>>res;
                std::unordered_set<tuple<int,int,int>>s;
                std::sort(nums);
                for (int i=0; i<nums.size()-2;i++){
                    for (int j=i+1; j<nums.size()-1;j++) {
                        for (int k=j+1; k<nums.size(); k++){
                            if (nums[i] +nums[j]+nums[k] == 0) {
                                if (s.find((nums[i],nums[j],nums[k])) == s.end()){
                                    res.push_back(std::vector<int>{nums[i], nums[j],nums[k]});
                                    s.insert((nums[i],nums[j],nums[k]));
                                }
                            }
                        }
                    }
                }
                return res;

            2.) Extending from 2sum 2
                - Sorting array helps us deal with repeats
                - for each element in nums, we run two sum 2 on the remaining subarray
                
                

        */
        std::vector<vector<int>>res;
        std::sort(nums.begin(), nums.end());
        int l,r, target;
        for (int i=0; i< nums.size()-2;i++){
            if (i > 0 && nums[i] == nums[i-1]) {
                continue;
            }
            l = i+1;
            r = nums.size()-1;
            target = 0 - nums[i];
            while (l<r) {
                if (l<r && nums[l]+nums[r] > target){
                    --r;
                }else if (l<r && nums[l]+nums[r] < target){
                    ++l;
                }else{
                    res.push_back({nums[i],nums[l],nums[r]});
                    ++l;
                    while (l<r && nums[l] == nums[l-1]) {
                        ++l;
                    }
                    --r;
                    while (l<r && nums[r] == nums[r+1]) {
                        --r;
                    }
                    

                }
            }
        }
        return res;
        
        
    }
};

File: 15_3Sum/solution.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        /*
            Input: integer array nums
            Output: list of list of integers on the triplets summing to 0 and they have a unique index in nums.

            Strategy 
            - sort, then for every pair of numbers in nums, binary search the remaining range for the number
            that will add up to 0.
            - TC: sort O(nlogn), binary search O(logn), nested loop O(n2) total: O(n2), SC: O(n)

        */
        
        
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        int target = 0;
        int left = 0;
        int mid = 0;
        int right = nums.length-1;
        int prev_i = Integer.MIN_VALUE;
        int prev_j = Integer.MIN_VALUE;
        Arrays.sort(nums);
        
        for(int i=0; i<nums.length-2; i++) {
            if (nums[i] == prev_i) continue;
            prev_i = nums[i];
            
            for (int j=i+1; j<nums.length-1; j++) {
                if (nums[j] == prev_j) continue;
                prev_j = nums[j];
                target = 0 - nums[i] - nums[j];
                left = j+1;
                right = nums.length-1;
                List<Integer> triplets = new ArrayList<Integer>();
                while (left <= right) {
                    mid = left + (right-left)/2;
                    if (nums[mid] == target) {
                        
                        triplets.add(nums[i]);
                        triplets.add(nums[j]);
                        triplets.add(nums[mid]);
                        res.add(triplets);
                        break;
                    }else if (nums[mid] < target) {
                        left = mid + 1;
                    }else{
                        right = mid-1;
                    }
                }
                
            }
        }
        return res;
    }
}
